apps/src/music/progress/MusicValidator.ts

```diff
@@ -414,4 +433,33 @@ export default class MusicValidator extends Validator {
   getValidationResults(): ValidationResult[] | undefined {
     return undefined;
   }
+
+  // Validates that both playback event arrays are equivalent based on id, type, and starting measure.
+  validatePlaybackEventsEquivalent(): boolean {
+    const studentEvents = [...this.getPlaybackEvents()];
+    const exemplarEvents = this.getExemplarPlaybackEvents();
+
+    if (studentEvents.length !== exemplarEvents.length) {
+      return false;
+    }
+
+    // For each exemplar event, find a matching student event.
+    return exemplarEvents.every(exemplar => {
+      const matchIndex = studentEvents.findIndex(
+        event =>
+          event.id === exemplar.id &&
+          event.type === exemplar.type &&
+          event.when === exemplar.when
+      );
+
+      if (matchIndex === -1) {
+        // No matching event found.
+        return false;
+      }
+
+      // Remove the matched event to prevent double matching.
+      studentEvents.splice(matchIndex, 1);
```

Comment:
Since we avoid emitting duplicate events to the timeline, is this needed?

Reply:
That could be! I wasn't sure at what point we prevented the duplicate, so this seemed like a reasonable safeguard. I'll manually test to confirm it's not needed it, and then likely remove it!

Reply:
I ended up re-working this after considering your feedback: https://github.com/code-dot-org/code-dot-org/pull/64114/commits/626a33b7f26085ed25859552a8d066621d4815aa

What I found is that the list of playback events does still include duplicates, even though we don't actually add them to the timeline. However, since the thing we care about is the final timeline, we just need to make sure all the student events can be found on the exemplar and vice versa. This technically makes things more permissive, since we just ignore the duplicate events in either array.